integration/swarm/leadership_test.go

```diff
@@ -0,0 +1,209 @@
+	names := map[string]string{}
+
+	defer func() {
+		for _, d := range daemons {
+			if d != nil {
+				d.Stop(t)
+			}
+		}
+	}()
+
+	// this closure expresses the name of a daemon as "daemons[i]" instead of
+	// the daemon ID, which makes the test logs easier to understand
+	name := func(d *daemon.Daemon) string {
+		return names[d.ID()]
+	}
+
+	// Create 3 daemons, all managers
+	daemons[0] = swarm.NewSwarm(t, testEnv, daemon.WithExperimental)
+	names[daemons[0].ID()] = "daemons[0]"
+	t.Logf("initial manager (%v) has ID: %v", name(daemons[0]), daemons[0].ID())
+
+	daemons[1] = daemon.New(t, daemon.WithExperimental, daemon.WithSwarmPort(daemon.DefaultSwarmPort+1))
+	daemons[1].StartAndSwarmJoin(t, daemons[0], true)
+	names[daemons[1].ID()] = "daemons[1]"
+	t.Logf("first joined manager (%v) has ID: %v", name(daemons[1]), daemons[1].ID())
+
+	daemons[2] = daemon.New(t, daemon.WithExperimental, daemon.WithSwarmPort(daemon.DefaultSwarmPort+2))
+	daemons[2].StartAndSwarmJoin(t, daemons[0], true)
+	names[daemons[2].ID()] = "daemons[2]"
+	t.Logf("second joined manager (%v) has ID: %v", name(daemons[2]), daemons[2].ID())
+
+	isLeader := func(d *daemon.Daemon, p poll.LogT) bool {
+		p.Logf("checking if %v is the leader", name(d))
+
+		// get the node directly, instead of through GetNode, so we
+		// can time out early if the attempt fails
+		cli := d.NewClientT(t)
+		defer cli.Close()
+
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+		defer cancel()
+
+		node, _, err := cli.NodeInspectWithRaw(ctx, d.NodeID())
+		if err != nil {
+			p.Logf("%v did not respond to node inspect: %v", name(d), err)
+			return false
+		}
+
+		return node.ManagerStatus.Leader
+	}
+
+	// find the first leader
+	var leader *daemon.Daemon
+	// these poll functions are all quite similar, but different enough that
+	// they're not coalesced into one big function.
+	poll.WaitOn(t, func(p poll.LogT) poll.Result {
+		// we're looking, in this poll, for 1 leader and 2 followers.
+		numLeaders := 0
+		numFollowers := 0
+		for _, d := range daemons {
+			if isLeader(d, p) {
+				leader = d
+				numLeaders++
+				p.Logf("%v is a leader", name(d))
+			} else {
+				numFollowers++
+				p.Logf("%v is not a leader", name(d))
+			}
+		}
+		if numLeaders == 0 {
+			return poll.Continue("no leader yet found")
+		}
+		if numLeaders > 1 {
+			// this case should not occur, but if it did, this would raise an
+			// immediate red flag and should totally stop the test.
+			return poll.Error(fmt.Errorf("more than one node claims to be the leader"))
+		}
+		if numFollowers != 2 {
+			return poll.Continue("number of followers is wrong, we expect 2 but have %v", numFollowers)
+		}
+		return poll.Success()
+	})
+	t.Logf("first leader is %v", name(leader))
+	assert.Check(
+		t, is.Equal(leader, daemons[0]),
+		"%v was the first node in the swarm, and adding nodes"+
+			"should not have caused a leader election, but %v is the leader",
+		name(daemons[0]), name(leader),
```

Comment:
I'm a bit on the fence on the `name()` function and use;

- looks like in any location where we use, we also have the index, so here we're doing `name(daemons[0])`, which returns `"daemons[0]"` (which would just as well be done in the format string; `daemon[%d]`
- the log-files are stored under the daemon's _ID_, so it's easier to have the `ID` when debugging, because then you can find the right logs (and other artefacts) in the `bundle.tgz`

Reply:
LOL it's really noisy though

```
16:21:32 --- PASS: TestSwarmLeaderElection (36.80s)
16:21:32     leadership_test.go:47: Creating a new daemon
16:21:32     daemon.go:336: [dc3be862f97eb] waiting for daemon to start
16:21:32     daemon.go:336: [dc3be862f97eb] waiting for daemon to start
16:21:32     daemon.go:364: [dc3be862f97eb] daemon started
16:21:32     leadership_test.go:49: initial manager (daemons[0]) has ID: dc3be862f97eb
16:21:32     leadership_test.go:51: Creating a new daemon
16:21:32     daemon.go:336: [dccd3ca7045f9] waiting for daemon to start
16:21:32     daemon.go:336: [dccd3ca7045f9] waiting for daemon to start
16:21:32     daemon.go:364: [dccd3ca7045f9] daemon started
16:21:32     leadership_test.go:54: first joined manager (daemons[1]) has ID: dccd3ca7045f9
16:21:32     leadership_test.go:56: Creating a new daemon
16:21:32     daemon.go:336: [d8ef7603fe7e4] waiting for daemon to start
16:21:32     daemon.go:336: [d8ef7603fe7e4] waiting for daemon to start
16:21:32     daemon.go:364: [d8ef7603fe7e4] daemon started
16:21:32     leadership_test.go:59: second joined manager (daemons[2]) has ID: d8ef7603fe7e4
16:21:32     leadership_test.go:62: checking if daemons[0] is the leader
16:21:32     leadership_test.go:93: daemons[0] is a leader
16:21:32     leadership_test.go:62: checking if daemons[1] is the leader
16:21:32     leadership_test.go:96: daemons[1] is not a leader
16:21:32     leadership_test.go:62: checking if daemons[2] is the leader
16:21:32     leadership_test.go:96: daemons[2] is not a leader
16:21:32     leadership_test.go:112: first leader is daemons[0]
16:21:32     daemon.go:472: [dc3be862f97eb] Stopping daemon
16:21:32     daemon.go:307: [dc3be862f97eb] exiting daemon
16:21:32     daemon.go:459: [dc3be862f97eb] Daemon stopped
16:21:32     leadership_test.go:121: stopped daemon daemons[0]
16:21:32     leadership_test.go:62: checking if daemons[1] is the leader
16:21:32     leadership_test.go:74: daemons[1] did not respond to node inspect: Cannot connect to the Docker daemon at unix:///tmp/docker-integration/dccd3ca7045f9.sock. Is the docker daemon running?
16:21:32     leadership_test.go:140: daemons[1] is not a leader
16:21:32     leadership_test.go:62: checking if daemons[2] is the leader
16:21:32     leadership_test.go:74: daemons[2] did not respond to node inspect: Cannot connect to the Docker daemon at unix:///tmp/docker-integration/d8ef7603fe7e4.sock. Is the docker daemon running?
16:21:32     leadership_test.go:140: daemons[2] is not a leader
16:21:32     leadership_test.go:62: checking if daemons[1] is the leader
16:21:32     leadership_test.go:74: daemons[1] did not respond to node inspect: Cannot connect to the Docker daemon at unix:///tmp/docker-integration/dccd3ca7045f9.sock. Is the docker daemon running?
16:21:32     leadership_test.go:140: daemons[1] is not a leader
16:21:32     leadership_test.go:62: checking if daemons[2] is the leader
16:21:32     leadership_test.go:74: daemons[2] did not respond to node inspect: Cannot connect to the Docker daemon at unix:///tmp/docker-integration/d8ef7603fe7e4.sock. Is the docker daemon running?
16:21:32     leadership_test.go:140: daemons[2] is not a leader
16:21:32     leadership_test.go:62: checking if daemons[1] is the leader
16:21:32     leadership_test.go:74: daemons[1] did not respond to node inspect: Cannot connect to the Docker daemon at unix:///tmp/docker-integration/dccd3ca7045f9.sock. Is the docker daemon running?
16:21:32     leadership_test.go:140: daemons[1] is not a leader
16:21:32     leadership_test.go:62: checking if daemons[2] is the leader
16:21:32     leadership_test.go:74: daemons[2] did not respond to node inspect: Cannot connect to the Docker daemon at unix:///tmp/docker-integration/d8ef7603fe7e4.sock. Is the docker daemon running?
16:21:32     leadership_test.go:140: daemons[2] is not a leader
16:21:32     leadership_test.go:62: checking if daemons[1] is the leader
16:21:32     leadership_test.go:74: daemons[1] did not respond to node inspect: Cannot connect to the Docker daemon at unix:///tmp/docker-integration/dccd3ca7045f9.sock. Is the docker daemon running?
16:21:32     leadership_test.go:140: daemons[1] is not a leader
16:21:32     leadership_test.go:62: checking if daemons[2] is the leader
16:21:32     leadership_test.go:74: daemons[2] did not respond to node inspect: Cannot connect to the Docker daemon at unix:///tmp/docker-integration/d8ef7603fe7e4.sock. Is the docker daemon running?
16:21:32     leadership_test.go:140: daemons[2] is not a leader
16:21:32     leadership_test.go:62: checking if daemons[1] is the leader
16:21:32     leadership_test.go:74: daemons[1] did not respond to node inspect: Cannot connect to the Docker daemon at unix:///tmp/docker-integration/dccd3ca7045f9.sock. Is the docker daemon running?
16:21:32     leadership_test.go:140: daemons[1] is not a leader
16:21:32     leadership_test.go:62: checking if daemons[2] is the leader
16:21:32     leadership_test.go:140: daemons[2] is not a leader
16:21:32     leadership_test.go:62: checking if daemons[1] is the leader
16:21:32     leadership_test.go:137: daemons[1] is a leader
16:21:32     leadership_test.go:62: checking if daemons[2] is the leader
16:21:32     leadership_test.go:140: daemons[2] is not a leader
16:21:32     leadership_test.go:157: leader after stopping the old leader is daemons[1]
16:21:32     daemon.go:336: [dc3be862f97eb] waiting for daemon to start
16:21:32     daemon.go:336: [dc3be862f97eb] waiting for daemon to start
16:21:32     daemon.go:336: [dc3be862f97eb] waiting for daemon to start
16:21:32     daemon.go:336: [dc3be862f97eb] waiting for daemon to start
16:21:32     daemon.go:364: [dc3be862f97eb] daemon started
16:21:32     leadership_test.go:62: checking if daemons[0] is the leader
16:21:32     leadership_test.go:62: checking if daemons[1] is the leader
16:21:32     leadership_test.go:62: checking if daemons[2] is the leader
16:21:32     daemon.go:472: [dc3be862f97eb] Stopping daemon
16:21:32     daemon.go:307: [dc3be862f97eb] exiting daemon
16:21:32     daemon.go:459: [dc3be862f97eb] Daemon stopped
16:21:32     daemon.go:472: [dccd3ca7045f9] Stopping daemon
16:21:32     daemon.go:307: [dccd3ca7045f9] exiting daemon
16:21:32     daemon.go:459: [dccd3ca7045f9] Daemon stopped
16:21:32     daemon.go:472: [d8ef7603fe7e4] Stopping daemon
16:21:32     daemon.go:307: [d8ef7603fe7e4] exiting daemon
16:21:32     daemon.go:459: [d8ef7603fe7e4] Daemon stopped
16:21:32 PASS
```

Reply:
It is insanely noisey. This is a problem, I realize, because we output _all_ test results, even if the test passes, which conflicts with the fact that having this much test output makes it way easier to see where the test went wrong (as was useful when I was debugging the test itself).

Maybe we should run CI without verbose mode, though? What good does verbose mode do besides creating more useless output?